scripts/ligandtocore/QD_functions.py

```diff
@@ -0,0 +1,357 @@
+
+    # searches for functional groups (defined by functional_group_list) within the ligand; duplicates are removed
+    get_match = ligand_rdkit.GetSubstructMatches
+    matches = [get_match(mol) for mol in functional_group_list]
+    matches = list(itertools.chain(*matches))
+
+    # rotates the functional group hydrogen atom in addition to returning the indices of various important ligand atoms
+    ligand_list = [copy.deepcopy(ligand) for i,match in enumerate(matches) if 
+                   match[1] != matches[i - 1][1] or i == 0]
+
+    [ligand.delete_atom(ligand[matches[i][0] + 1]) for i,ligand in enumerate(ligand_list)]
+    matches = [item[1] for item in matches]
+
+    if not ligand_list:
+        print('No functional groups were found for ' + str(ligand.get_formula()))
+
+    return ligand_list, matches
+
+
+def rotate_ligand(core, ligand, core_index, ligand_index):
+    """
+    Connects two molecules by alligning the vectors of two bonds
+    """
+    ligand = copy.deepcopy(ligand)
+
+    # Defines first atom on coordinate list (hydrogen), atom connected to it and vector representing bond between them
+    core_at1 = core[core_index]         # core dummy atom
+    core_at2 = core[-1]                 # core center of mass
+    core_vector = core_at1.vector_to(core_at2)
+    lig_at1 = ligand[ligand_index + 1]  # ligand heteroatom
+    lig_at2 = ligand[-1]                # ligand center of mass
+    lig_vector = lig_at2.vector_to(lig_at1)
+
+    # Rotation of ligand - aligning the ligand and core vectors
+    rotmat = rotation_matrix(lig_vector, core_vector)
+    ligand.rotate(rotmat)
+    ligand.translate(lig_at1.vector_to(core_at1))
+
+    # Translation of the ligand
+    hc_vec = lig_at1.vector_to(core_at1)
+    ligand.translate(hc_vec)
+
+    # Deletes the core dummy atom and ligand center of mass
+    ligand.delete_atom(lig_at2)
+    core.delete_atom(core_at1)    
+
+    return ligand, lig_at1
+
+
+# Function for matrix rotation
+def rotation_matrix(vec1, vec2):
+    """
+    Calculates the rotation matrix rotating *vec1* to *vec2*. Vectors can be any containers with 3 numerical values. They don't need to be normalized. Returns 3x3 numpy array.
+    """
+    a = np.array(vec1) / np.linalg.norm(vec1)
+    b = np.array(vec2) / np.linalg.norm(vec2)
+    v1, v2, v3 = np.cross(a, b)
+    M = np.array([[0, -v3, v2], [v3, 0, -v1], [-v2, v1, 0]])
+
+    return (np.identity(3) + M + np.dot(M, M)/(1+np.dot(a, b)))
+
+
+def combine_QD(core, ligand_list):
+    """
+    combine the rotated ligands with the core, creating a bond bewteen the core and ligand in the process
+    """
+    QD = copy.deepcopy(core)
+
+    # create a list of ligand atoms and intraligand bonds
+    ligand_bonds = np.concatenate([ligand.bonds for ligand in ligand_list])
+    ligand_atoms = np.concatenate(ligand_list)
+
+    # Combined the ligand bond and atom list with the core
+    [QD.add_atom(atom) for atom in ligand_atoms]
```

Comment:
Do not use list comprehension if you are not using the resulting list, instead use a simple `for` loop